## Option Pricing using Monte Carlo Simulation

Option pricing using Monte Carlo simulation and geometric brownian motion. Compared with Black Scholes method.

Will calculate CALL and PUT Payoffs.

In [1]:
# Libraries

import numpy as np
from scipy.stats import norm

In [2]:
# Variables:

S = 385                 # Spot price
K = 350                 # Strike price
T = 0.083               # Time to expiration (Years)
r = 0.04                # Discount rate (%)     
sigma = 0.15            # Volatility of stock(%)


# Variables for Monte Carlo:

steps = 30              # Number of days option runs for
simulations = 5         # Number of simulations 

In [3]:
# Brownian motion for daily prices:
dt = T / steps

In [4]:
# Drift:
drift = (r-(sigma**2)/2) * dt

In [5]:
# Time:
a = sigma * np.sqrt(dt)

In [6]:
# Random variable: Creating random variables for each day of stock.

# Using standard normal distribution with mean=0, variance=1, (Nsimulation for Nsteps) 
x = np.random.normal(0,1,(simulations, steps))      
print(x)

[[-2.10418211e-01 -6.21302362e-01 -2.13191212e-01 -4.62560375e-01
  -2.20605956e+00  5.26344419e-01  4.00916155e-01  1.73949968e+00
  -4.30734248e-01 -1.19431322e-02 -1.09031959e+00 -7.74083258e-01
   1.66610691e+00 -5.94791235e-01  1.43373860e+00 -8.84225149e-01
   1.07093844e+00 -8.35916748e-01 -6.82328287e-01 -7.88431315e-01
   3.64156795e-01  1.06885458e+00 -6.40627039e-01 -7.12757190e-03
  -4.29684832e-01  1.59470576e+00  1.12779345e+00  1.26619975e+00
  -9.02315543e-01 -1.01273385e-01]
 [-9.03642106e-01  2.94289291e-01 -1.23687273e+00 -4.78751698e-01
  -9.95767457e-01  7.99347684e-01 -8.73872577e-01  2.30739990e+00
   3.50986715e-01  1.99738211e+00  1.03170095e+00 -2.06371161e+00
  -1.16860169e+00  6.28378220e-01 -4.71034116e-01  5.03807312e-01
  -1.55030406e+00 -7.68035681e-01 -1.48389323e+00  1.49977331e+00
   7.78681274e-02 -2.80008437e+00  6.84572603e-01  1.13180903e+00
  -8.84108833e-01  1.70260229e+00 -9.05678533e-02  8.63074644e-01
   8.12784695e-01  4.96791480e-01]
 [ 2.1

Above we can see that for each of the 5 arrays (simulations), there are 30 elements (steps). Each element represents one day.

In [7]:
# Matrix of stock price values

Smat = np.zeros((simulations, steps))       # Matrix initially loaded with zero values
Smat[:,0]+= S                               # put 385 (Spot price) on the first element of each of the 5 arrays
Smat

array([[385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [385.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0., 

Above we can see that the each of the 5 arrays contain a value of 385 (Spot price) at the 1st element, with 0's afterwards. These 0's will be exchanged for a value based on the random variables generated prevoiusly.

In [9]:
# Brownian Motion: (from the 2nd element onwards, we have the Brownian motion values)
# Each simulation represents a different Brownian Motion path.

for i in range(1, steps):
    Smat[:,i] += Smat[:,i-1] * np.exp(drift + a*x[:,i])
Smat

array([[385.        , 383.14782585, 382.53431987, 381.17110784,
        374.62382965, 376.21272097, 377.43465238, 382.68087224,
        381.4129002 , 381.40729803, 378.170402  , 375.89769581,
        380.90192382, 379.14876285, 383.49254858, 380.85674746,
        384.11900633, 381.62432667, 379.60557341, 377.28152992,
        378.3971717 , 381.63208829, 379.73821387, 379.74706418,
        378.4919468 , 383.31471193, 386.77148316, 390.68582682,
        387.94520954, 387.66618651],
       [385.        , 385.92566735, 382.20822805, 380.7975291 ,
        377.84757862, 380.2683439 , 377.68555146, 384.65491006,
        385.75226965, 391.91069528, 395.14529018, 388.79439341,
        385.25678349, 387.20235761, 385.79671646, 387.36411013,
        382.68528941, 380.40360028, 376.005818  , 380.51179296,
        380.77592584, 372.48558148, 374.53267845, 377.9222257 ,
        375.32505467, 380.43118625, 380.18967901, 382.81788086,
        385.31133659, 386.8553437 ],
       [385.        , 387.3972

Above we can see that we have 5 different Brownian Motion paths (simulations) for the stock price.

### Calulate Payoffs for CALL and PUT:

In [10]:
# CALL payoff values:

c = (Smat[:,-1]) - K        # Call = 'last value of Smat' subtract 'Strike Price'

for i in range(len(c)):
    if c[i] < 0:
        c[i] = 0            # where values are less than 1, make these values equal zero
    else:
        c[i] = c[i]         # otherwise take value as it is                
c

array([37.66618651, 36.8553437 , 33.93789417, 31.41225523, 23.14993703])

In [11]:
# PUT payoff values:

p = K - (Smat[:,-1])        # Put = 'Strike Price' subtract the 'Spot price' 
for i in range(len(p)):
    if p[i] < 0:
        p[i] = 0            # where values are less than 1, make all values equal zero
    else:
        p[i] = p[i]         # otherwise take value as it is                
p

array([0., 0., 0., 0., 0.])

In [12]:
# Take the average of payoff simulations:

call_payoff = np.mean(c)
print('Call payoff =', call_payoff)

put_payoff = np.mean(p)
print('Put payoff =', put_payoff)

Call payoff = 32.60432332862446
Put payoff = 0.0


### Discount price back to present day:

In [13]:
call = call_payoff*np.exp(-r*T)
put = put_payoff*np.exp(-r*T)

In [14]:
print('Call payoff (Discounted to present day) =', call)
print('Put payoff (Discounted to present day) =', put)

Call payoff (Discounted to present day) = 32.496256465429326
Put payoff (Discounted to present day) = 0.0


We can see that the call payoff has value of 32.49, whereas the put payoff has a value of 0.0. This is because for these given spot and strike prices they only work in favor of a call option. Hence a Put option at these prices would not be executed. 

## Black Scholes Method:

In [15]:
# Black Scholes d1 & d2:

d1 = (np.log(S/K) + (r + sigma**2/2)*T) / sigma*np.sqrt(T)
d2 = (np.log(S/K) + (r - sigma**2/2)*T) / sigma*np.sqrt(T)

In [16]:
# Black Scholes Call option price:

call_bs = (norm.cdf(d1,0,1) * S) - (norm.cdf(d2,0,1) * K*np.exp(-r*T))
call_bs

21.31220748950824

In [17]:
call_discounted = call_bs*np.exp(-r*T)
print('Call payoff (Discounted to present day) =', call_discounted)

Call payoff (Discounted to present day) = 21.24156828660434


Using the same variables, the values have have been entered into the Black Scholes model. The call payoff has a value of 21.24.